### PR TITLE
WIP: PCI segment one hotplug support

### PR DESCRIPTION
**Pull request by liujing2**
Hi @mcastelino,
I tried the hotplug on PCI segment 1 today, and this is still in progress.

Current status is,
when hotplugging onto 1.pcie.0, qemu got corrent response to send
GED event but kernel shows nothing in dmesg.
Need some debug then.

### CODE REVIEW

**Review by mcastelino — COMMENTED** [review]
@liujing2  can we call both the methods in the same interrupt routine. Will that work. i.e call
\\_SB.PCI2.PCNT and \\_SB.PCI0.PCNT

File: hw/acpi/aml-build.c [hunk 1]
```diff
@@ -41,7 +41,10 @@
 #define PCI_HOST_BRIDGE_IO_0_MIN_ADDR      0x0000
 #define PCI_HOST_BRIDGE_IO_0_MAX_ADDR      0x0cf7
 #define PCI_HOST_BRIDGE_IO_1_MIN_ADDR      0x0d00
-#define PCI_HOST_BRIDGE_IO_1_MAX_ADDR      0xffff
+#define PCI_HOST_BRIDGE_IO_1_MAX_ADDR      0x9fff
```
Comment (mcastelino) [inline hw/acpi/aml-build.c:44]: @liujing2 longer term we should decide if these constants should even be in this file. Ideally they should be driven from the platform definition (in our case virt). So piix and q35 and virt should define these values in a platform specific file.

File: hw/i386/virt/acpi.c [hunk 2]
```diff
@@ -48,6 +48,9 @@ typedef struct VirtAcpiState {
     AcpiPciHpState pcihp_state;
```
Comment (mcastelino) [inline hw/i386/virt/acpi.c:48]: @liujing2 we should make this a list longer term.

File: hw/acpi/aml-build.c (hunk 1 above)
Comment (liujing2) [reply hw/acpi/aml-build.c:44]: @mcastelino If only virt uses multi-segment PCI, agree with you. piix and q35 own the whole MIN~MAX ADDR and virt separates them for each segment.

File: hw/i386/virt/acpi.c (hunk 2 above)
Comment (liujing2) [reply hw/i386/virt/acpi.c:48]: @mcastelino Yes. One problem about pcihp_state is how to distinguish which pcihp_state[i] to give in [virt_device_plug_cb](https://github.com/intel/nemu/blob/857d652a0faccfbf61d33ac4efbca8b1581c9bae/hw/i386/virt/acpi.c#L80), and this seems similar to the problem of PCNT.

**Comment by liujing2** [comment]
@mcastelino 
Mark the experiments as below.

(1) I hardcode like this, then hotplug to PCI0 works but PCI2(1.pcie.0) doesn't. On the contrary, PCI2 works but PCI0 doesn't. 
```
 case GED_PCI_HOTPLUG:
        method = aml_acquire(aml_name("\\_SB.PCI2.BLCK"), 0xFFFF); //put PCI2 phase first, but PCI0 works.
        aml_append(method, aml_call0("\\_SB.PCI2.PCNT"));
        aml_append(method, aml_release(aml_name("\\_SB.PCI2.BLCK")));
        method = aml_acquire(aml_name("\\_SB.PCI0.BLCK"), 0xFFFF);
        aml_append(method, aml_call0("\\_SB.PCI0.PCNT"));
        aml_append(method, aml_release(aml_name("\\_SB.PCI0.BLCK")));
```
(2)  I hardcode like this, only hotplug on the first PCI works. In such case, PCI2 works.
```
    case GED_PCI_HOTPLUG:
        /* Take the PCI lock and trigger a PCI rescan */
        method = aml_acquire(aml_name("\\_SB.PCI0.BLCK"), 0xFFFF);
        aml_append(method, aml_call0("\\_SB.PCI2.PCNT")); //The first, works.
        aml_append(method, aml_call0("\\_SB.PCI0.PCNT"));
        aml_append(method, aml_release(aml_name("\\_SB.PCI0.BLCK")));
```

**Comment by mcastelino** [comment]
@liujing2 thanks for that experiment. Let me figure why it did not work. Worst case I can implement a top level method to demultiplex.

**Pull request merged**